shells/pipes-shell/source/pipe.js

```diff
@@ -50,9 +52,74 @@ const populateDispatcher = (dispatcher, composerFactory, storage, context) => {
     pec: async (msg, tid, bus) => {
       return await pec(msg, tid, bus);
     },
+    ingest: async (msg, tid, bus) => {
+      return await ingest(msg.entity, tid, bus);
+    },
     spawn: async (msg, tid, bus) => {
-      return await spawn(msg, tid, bus, composerFactory, storage, context);
+      const arc = await spawn(msg, tid, bus, composerFactory, storage, context);
+      return arc;
+    },
+    recipe: async (msg, tid, bus) => {
+      const arc = await bus.getAsyncValue(msg.tid);
+      if (arc) {
+        return await instantiateRecipeByName(arc, msg.recipe);
+      }
+    },
+    event: async (msg, tid, bus) => {
+      // find the arc from the tid in the message (not the tid for this request)
+      const arc = await bus.getAsyncValue(msg.tid);
+      if (arc) {
+        // find the particle from the pid in the message
+        const particle = arc.activeRecipe.particles.find(
+          particle => String(particle.id) === msg.pid
+        );
+        if (particle) {
+          log('firing PEC event for', particle.name);
+          // TODO(sjmiles): we need `arc` and `particle` here even though
+          // the two are bound together, figure out how to simplify
+          arc.pec.sendEvent(particle, /*slotName*/'', msg.eventlet);
+        }
+      }
     }
   });
   return dispatcher;
 };
+
+const composerFactory = (modality, bus, tid) => {
+  switch (modality) {
+    case 'ram': {
+      return new RamSlotComposer();
+    }
+    default: {
+      const composer = new RamSlotComposer({
+        top: 'top',
+        root: 'root',
+        modal: 'modal'
+      });
+      // TODO(sjmiles): hack in transaction identity, make this cleaner
+      composer.tid = tid;
+      // TODO(sjmiles): slotObserver could be late attached
+      // or we could attach a thunk that dispatches to an actual
+      // broker configured elsewhere.
+      composer.slotObserver = brokerFactory(bus);
+      return composer;
+    }
+  }
+};
+
+// `slot-composer` delegates ui work to a `ui-broker`
+const brokerFactory = bus => {
```

Comment:
This PR combines many changes that are not directly related to each other or the UI broker evolution. 
Would it be possible to split them apart? Would it be ok with you, if I port the `composerFactory` and `brokerFactory` changes alone from this file into the https://github.com/PolymerLabs/arcs/pull/3543, so that I can make Java pec work with new rendering e2e? Thanks.

Reply:
I don't understand what you are asking. I'm asking to pull these changes to get my work done, I believe this code is all related. 

Are you trying to avoid some of these changes?

Reply:
I don't think the changes to the `ingest` and `recipe` APIs, and the functionality in`sharing.js` are related to UI broker work. 
I suspect some of the changes in this PR might be here unintentionally (`process.js` and the couple lines in `arc.ts`). 
And it is also really hard for me to review a PR of this size.

I understand that you're ooo and don't have time to split the PR. You felt the pressure to send this PR out because you didn't like the changes I made in https://github.com/PolymerLabs/arcs/pull/3543/files#diff-8c5bb89e542af8354641e178ee21877dR33. This is why I suggested to port the subset of changes from this PR that will allow Java PEC to use the new rendering system. 

Please, let me know what you think.